### DISCUSSION TITLE
using ZMQ to pump information around Genie cloud

### DISCUSSION THREAD

**Discussion by j215chic** [GenieFramework/Genie.jl#644] [Q&A] [answered]
Hi there
this question is specific to GenieCloud.  We want to stream information into our web page, make decisions based on the streaming data and pipe the results out to another web page using ZMQ.   

We understand that there is a features_peerinfo the Configuration module which should return the ip address of the web page and we "THINK" we can use that for this scenario.  Here is the usecase:

we generate streaming data from an app and send it to port 5555 ( gotten from Web_page1. features_peerinfo)  Web_page1 starts up Web_page2 which ( we assume) gets assigned a port by the system (lets say it gets 5556)  and Web_page1 queries Web_page2.features_peerinfo to get Web_page1.output port. This would result in

5555 => Web_page1 => 5556 => Web_page2.

we tend to use matrices to handle queuing routing so if there is a better way to do this please advise. If possible we'd like to start the web app and have all the ip addresses in one place so we can build out our environment in a flexible manner.  This would allow us to handle the situation where IP addresses and ports are not statically assigned.

**Comment by essenciary** [comment]
@j215chic I'm afraid I don't have a good enough understanding of the context of your app to comment on the flow. But what I can confirm is that features_peerinfo should provide the IP information from the request. However, as it's pretty rarely used, let us know if it doesn't work as expected.

**Reply by j215chic** [reply]
@essenciary  thank you for the reply 

fair enough.  How about this.

we need to send messages, via ZMQ, directly into our GenieCloud hosted web app from a REMOTE system. To do this we need the IP address and PORT number of the GenieCloud app.  

Are you assigning STATIC ip's and can we define our port numbers?

We cannot use [JuliaWebAPI]( https://github.com/JuliaWeb/JuliaWebAPI.jl/blob/master/README.md  ) as we are trying to avoid adding PKG uncertainty.

**Reply by essenciary** [reply]
You can't use the IP directly as it's not exposed -- so you need to use the app's host instead. As for ports, we only expose the "web" ports (80 and 443) which are mapped to port 8000 of the Genie app. I hope this helps.

**Reply by essenciary** [reply]
Why don't you connect the GenieCloud app to ZMQ via something like https://github.com/JuliaInterop/ZMQ.jl

**Reply by j215chic** [reply]
Hi there
we use ZMQ all the time I think I mentioned that in the OP. 

we need to know the IP address and port number that it's started with. The **REMOTE** system  sending the information needs to know where to send it to.  

From your answer it seems we can't as the IP isn't published.

**Reply by essenciary** [reply, marked as answer]
Correct, the IP of the GenieCloud app is not directly accessible, you'd be getting the IP of the load balancer. You'd need to use the host.

**Reply by j215chic** [reply]
ok thank you. We wanted to check to make sure.
